sbkeys/generate-aws-sbkeys

```diff
@@ -0,0 +1,335 @@
+mkdir -p "${OUTPUT_DIR}"
+
+if [ -n "${SDK_IMAGE:-}" ] ; then
+  docker run -a stdin -a stdout -a stderr --rm \
+    --user "$(id -u):$(id -g)" \
+    --security-opt label:disable \
+    -v "${OUTPUT_DIR}":/tmp/output \
+    -v "${SBKEYS_SCRIPT}":/tmp/sbkeys \
+    -v "${AWS_KMS_PKCS11_CONF}":/tmp/aws-kms-pkcs11-conf \
+    ${AWS_ACCESS_KEY_ID:+-e AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID} \
+    ${AWS_SECRET_ACCESS_KEY:+-e AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY} \
+    ${AWS_SESSION_TOKEN:+-e AWS_SESSION_TOKEN=$AWS_SESSION_TOKEN} \
+    -e AWS_REGION="${AWS_REGION}" \
+    -e AWS_DEFAULT_REGION="${AWS_REGION}" \
+    -e PK_CA="${PK_CA}" \
+    -e KEK_CA="${KEK_CA}" \
+    -e DB_CA="${DB_CA}" \
+    -e VENDOR_CA="${VENDOR_CA}" \
+    -e SHIM_SIGN_KEY="${SHIM_SIGN_KEY}" \
+    -e CODE_SIGN_KEY="${CODE_SIGN_KEY}" \
+    -e CONFIG_SIGN_KEY="${CONFIG_SIGN_KEY}" \
+    -e AWS_KMS_PKCS11_CONF="/tmp/aws-kms-pkcs11-conf" \
+    -e OUTPUT_DIR="/tmp/output" \
+    -w /tmp \
+    "${SDK_IMAGE}" bash /tmp/sbkeys
+else
+  export PK_CA KEK_CA DB_CA VENDOR_CA
+  export CODE_SIGN_KEY CONFIG_SIGN_KEY
+  export AWS_REGION AWS_KMS_PKCS11_CONF OUTPUT_DIR
+  bash "${SBKEYS_SCRIPT}"
+fi
+
+exit
+
+# =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+set -euo pipefail
+
+WORKDIR="$(mktemp -d)"
+cd "${WORKDIR}"
+cleanup() {
+  rm -rf "${WORKDIR}"
+}
+trap 'cleanup' EXIT
+
+export XDG_CONFIG_HOME="${WORKDIR}/.config"
+mkdir -p "${XDG_CONFIG_HOME}/aws-kms-pkcs11"
+cp "${AWS_KMS_PKCS11_CONF}" "${XDG_CONFIG_HOME}/aws-kms-pkcs11/config.json"
+
+export AWS_DEFAULT_OUTPUT="text"
+export AWS_KMS_PKCS11_DEBUG=1
+export PKCS11_MODULE_PATH="/usr/lib64/pkcs11/aws_kms_pkcs11.so"
+
+# Fetch CA certificates.
+getcacert() {
+  local arn ca
+  arn="${1:?}"
+  ca="${2:?}"
+  aws acm-pca get-certificate-authority-certificate \
+    --certificate-authority-arn "${arn}" \
+    --query 'Certificate' > "${ca}.crt"
+}
+
+getcacert "${PK_CA}" "PK"
+getcacert "${KEK_CA}" "KEK"
+getcacert "${DB_CA}" "db"
+getcacert "${VENDOR_CA}" "vendor"
+
+# Add X.509 extension for code signing.
+cat <<'EOF' > codesign.json
+{
+  "Extensions": {
+    "ExtendedKeyUsage": [
+      {
+        "ExtendedKeyUsageType": "CODE_SIGNING"
+      },
+      {
+        "ExtendedKeyUsageObjectIdentifier": "1.3.6.1.4.1.311.10.3.6"
+      }
+    ]
+  }
+}
+EOF
+
+gencert() {
+  local key token cn ca_arn cert_arn
+  key="${1:?}"
+  token="${2:?}"
+  cn="${3:?}"
+  ca_arn="${4:?}"
+
+  openssl req -new \
+    -key "pkcs11:token=${token}" -keyform engine -engine pkcs11 \
+    -subj "/CN=${cn}/" \
+    -out "${key}.csr"
+    cert_arn="$(\
+    aws acm-pca issue-certificate \
+      --certificate-authority-arn "${ca_arn}" \
+      --template-arn arn:aws:acm-pca:::template/BlankEndEntityCertificate_APICSRPassthrough/V1 \
+      --csr "fileb://${key}.csr" \
+      --api-passthrough "file://codesign.json" \
+      --signing-algorithm "SHA256WITHRSA" \
+      --validity Value=5,Type="YEARS" \
```

Comment:
> I assume the impact of those certificates expiring is the boot failing.

I expect it's firmware-specific, but QEMU seems happy with an expired certificate, if a VM is started with `-rtc base=2036-11-11T11:11:11`, and shim says ["UEFI explicitly allows expired certificates"](https://github.com/rhboot/shim/blob/e24681211ce8a71583b5233084bd1290c3c7e872/Cryptlib/Pk/CryptPkcs7Verify.c#L103).

In most cases it seems undesirable to have boot fail just because the system clock is wrong, like if the RTC battery dies and it resets back to the epoch.

> From an operational point of view, will there be a new set of keys per variant? Five years of lifetime seems safe now, as variants would phase out before that (perhaps aws-ecs-1 could come close). Should this be documented anyway?

For our builds, the plan is to create a new generation of signing keys whenever we introduce a newer kernel. Not sure where we'd document that though; we don't document our (annual) TUF root rotation schedule either.

Reply:
If I try to sign with an expired certificate then `pesign` complains:
```
  #14 9.402 Peer's Certificate has expired.
  #14 9.402 Searching db db.cer
  #14 9.402 Searching db db.cer
  #14 9.402 Signature has impossible time constraint: 1688764933 <= 1688763121
  #14 9.402 pesigcheck: "bootx64.efi" is invalid.
```

So the main operational impact should just be that we won't be able to (easily) use the same signing certs for more than five years, which seems fine. There's a balance to strike between not using the same certs forever vs. using few enough that it's reasonable to revoke based on the cert rather than a list of individual hashes.

Reply:
> > I assume the impact of those certificates expiring is the boot failing.
> 
> I expect it's firmware-specific, but QEMU seems happy with an expired certificate, if a VM is started with `-rtc base=2036-11-11T11:11:11`, and shim says ["UEFI explicitly allows expired certificates"](https://github.com/rhboot/shim/blob/e24681211ce8a71583b5233084bd1290c3c7e872/Cryptlib/Pk/CryptPkcs7Verify.c#L103).

That's interesting, thanks for digging it up!



> > From an operational point of view, will there be a new set of keys per variant? Five years of lifetime seems safe now, as variants would phase out before that (perhaps aws-ecs-1 could come close). Should this be documented anyway?
> 
> For our builds, the plan is to create a new generation of signing keys whenever we introduce a newer kernel. Not sure where we'd document that though; we don't document our (annual) TUF root rotation schedule either.

Viewed that way, this makes sense to me. Key rotation is a matter that concerns image providers. The mechanism exists to implement various policies. Bottlerocket doesn't need to prescribe (and document) a particular one.